portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -11,22 +11,111 @@
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
 /** Servlet that returns some example content. TODO: modify this file to handle comments data */
+
+
+
+
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
-
+  final int defaultdisplayNumber = 10;
+  final String commentEntityType = "Comment"; 
+  final String statementProperty = "statement";
+  final String timestampProperty = "timestamp";
+  final String userProperty = "user";
+  int displayNumber = defaultdisplayNumber;
+  UserService userService = UserServiceFactory.getUserService();
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    response.setContentType("text/html;");
-    response.getWriter().println("<h1>Hello world!</h1>");
+    Query query = new Query(commentEntityType).addSort(timestampProperty, SortDirection.DESCENDING);
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery  results = datastore.prepare(query);
+    
+    int countComments = 0;
+    List<Comment> comments = new ArrayList<>();
+    for (Entity entity : results.asIterable()){
+      if(displayNumber == countComments){
+        break;
+      }
+      countComments++;
+      String statement = (String)entity.getProperty(statementProperty);
+      long timestamp = (long)entity.getProperty(timestampProperty);
+      String user = (String)entity.getProperty(userProperty);
+      Comment newComment = new Comment(statement, timestamp, user);
+      comments.add(newComment);
+    }
+
+    response.setContentType("application/json;");
+    if(!userService.isUserLoggedIn()){
+      return;
+    }
+    Gson gson = new Gson();
+    String json = gson.toJson(comments);
+    response.getWriter().println(json);
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+
+    displayNumber = getNumberOfComments(request);
+    if(displayNumber == -1){
+      displayNumber = defaultdisplayNumber;
+    }
+    String comment = getComment(request);
+    //if comment is empty or if user is not logged in
+    if(comment.length() == 0 || !userService.isUserLoggedIn()){
+      response.sendRedirect("/index.html");
+      return;
+    }
+    
+    Entity commentEntity = new Entity(commentEntityType);
+    commentEntity.setProperty(statementProperty, comment);
+    long timestamp = System.currentTimeMillis();
+    commentEntity.setProperty(timestampProperty, timestamp);
+    String user = userService.getCurrentUser().getEmail();
+    commentEntity.setProperty(userProperty, user);
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
```

Comment:
Always catch exception, in case datastore service is not available.
Put a try catch if possible.